// auth0-hooks-templates/Gruntfile.js
const fs   = require('fs');
const path = require('path');

module.exports = function exec(grunt) {
  const TMPL_PATH = 'hooks/assets';
  const DEV_PATH  = 'hooks/assets/develop';

  const pkg  = require('./package');

  const minor = pkg.version.replace(/\.(\d)*$/, '');
  const major = pkg.version.replace(/\.(\d)*\.(\d)*$/, '');

  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    aws_s3: {
      options: {
        accessKeyId:     process.env.S3_KEY,
        secretAccessKey: process.env.S3_SECRET,
        bucket:          process.env.S3_BUCKET,
        region:          process.env.S3_REGION,
        uploadConcurrency: 5,
        params: {
          CacheControl: 'public, max-age=300'
        },
      },
      clean: {
        files: [
          { action: 'delete', dest: `${TMPL_PATH}/hooks.json` }
        ]
      },
      publish: {
        files: [
          {
            expand: true,
            cwd: 'release/',
            src: ['hooks.json'],
            dest: `${TMPL_PATH}`
          }
        ]
      },
      clean_dev: {
        files: [
          { action: 'delete', dest: `${DEV_PATH}/hooks.json` }
        ]
      },
      publish_dev: {
        files: [
          {
            expand: true,
            cwd: 'release/',
            src: ['hooks.json'],
            dest: `${DEV_PATH}`
          }
        ]
      },
    },
    http: {
      purge_templates: {
        options: {
          url: `${process.env.CDN_ROOT}/${TMPL_PATH}/hooks.json`,
          method: 'DELETE'
        }
      },
      purge_templates_dev: {
        options: {
          url: `${process.env.CDN_ROOT}/${DEV_PATH}/hooks.json`,
          method: 'DELETE'
        }
      }
    }
  });

  grunt.loadNpmTasks('grunt-contrib-copy');
  grunt.loadNpmTasks('grunt-aws-s3');
  grunt.loadNpmTasks('grunt-http');

  grunt.registerTask('templates', () => {
    const templates = { hooks: [] };

    fs.readdirSync(path.join('templates')).filter((file) => {
      return file.indexOf('.yaml') >= 0;
    }).map((tmpl) => {
      const template = grunt.file.readYAML(path.join('templates', tmpl));
      templates.hooks.push(template);
    });

    grunt.file.write(path.join('release', 'hooks.json'), JSON.stringify(templates, null, 2));
  });

  grunt.registerTask('purge-cdn', ['http:purge_templates']);

  grunt.registerTask('cdn', [
    'templates',
    'aws_s3:clean',
    'aws_s3:publish',
    'purge-cdn'
  ]);

  grunt.registerTask('purge-cdn-dev', ['http:purge_templates_dev']);

  grunt.registerTask('cdn-dev', [
    'templates',
    'aws_s3:clean_dev',
    'aws_s3:publish_dev',
    'purge-cdn-dev'
  ]);
};
